### ISSUE TITLE
Our python builds stopped working again

### ISSUE THREAD

**Issue by StingyJack** [microsoft/azure-pipelines-image-generation#1467]
The same build def and commit that worked last week do not work today. 

I have a build task to use the version of python needed (3.6.9), but that is apparently a fools task...

```
Version spec 3.6.9 for architecture x64 did not match any version in Agent.ToolsDirectory.
Versions in /opt/hostedtoolcache:
2.7.17 (x64)
3.5.9 (x64)
3.6.10 (x64)
3.7.6 (x64)
3.8.1 (x64)
If this is a Microsoft-hosted agent, check that this image supports side-by-side versions of Python at https://aka.ms/hosted-agent-software.
If this is a self-hosted agent, see how to configure side-by-side Python versions at https://go.microsoft.com/fwlink/?linkid=871498.
```

If you are really trying to improve in the area of builds and software maintenance, this is the place to start. Solve the problem in front of us first.

**Comment by vmapetr** [comment]
Hi @StingyJack! 

Thank you for reporting the issue. During recent image rollout we have updated Hostedtoolcache Python version from `3.6.9` to `3.6.10`.
Please check the documentation to find new updates:
Ubuntu 18.04 - https://github.com/actions/virtual-environments/blob/releases/ubuntu18/20200113/images/linux/Ubuntu1804-README.md
Ubuntu 16.04 - https://github.com/actions/virtual-environments/blob/releases/ubuntu16/20200113/images/linux/Ubuntu1604-README.md

To avoid `use Python version` task failures, you should use task without specified build version, e.g. `3.6` instead of `3.6.9`.

Example:
**VSTS - UsePythonVersion:**
```
  - task: UsePythonVersion@0
    displayName: 'Use Python $(VERSION)'
    inputs:
      versionSpec: '3.6'
      architecture: 'x64'
```
**GitHub - setup-python:**
```
- uses: actions/setup-python@v1
  with:
    python-version: '3.6' 
    architecture: 'x64'
```

With this approach currently installed build version will be used:

```
Found tool in cache: Python 3.6.10 x64
Prepending PATH environment variable with directory: /opt/hostedtoolcache/Python/3.6.10/x64
Prepending PATH environment variable with directory: /opt/hostedtoolcache/Python/3.6.10/x64/bin
```

**Comment by StingyJack** [comment]
We have screenshots of the docs pages on the day this broke. It did not list 3.6.10.  
![image](https://user-images.githubusercontent.com/15381181/73141863-9aff2900-4056-11ea-8985-6ebba6cc6b5d.png)

I'm not sure _when_ it was actually updated and made available to the public but linking me to the commits for a few readme files is not the same as when the published versions of those docs are available. Those published versions need to lead the actual updates by at least a week or two, and affected users should be getting some kind of notice that updates to the images are happening since it seems to be random.

Its not a new problem either. The docs did not get updated to list 3.6.9 for several weeks after it had been updated. and had broken builds unexpectedly, causing a late afternoon panic when I had another task on a deadline already.  [Other packages like NodeJs has been bumped two major versions before and the docs not been updated for > 24 days](https://github.com/microsoft/azure-pipelines-image-generation/issues/1393#issuecomment-561980746)

I hope you do realize that python (and many other programs) do not follow semantic versioning with releases. They may be compatible, but as a **service provider** you cannot be interrupting the service we are paying for without some kind of notice or assistance. Forcing me to upgrade is not assistance, it is forcing me to have my application(s) fully regression tested, and that is something that even Microsoft finds too costly to perform, so I dont know what makes it OK to foist it upon its customers.

**Comment by StingyJack** [comment]
@alepauly - did this get moved so you all would actually address it, or was this another attempt to relieve the debt you have/are accruing at the users expense?

**Comment by alepauly** [comment]
Hi @StingyJack, see https://github.com/actions/virtual-environments/issues/281. While that issue talks about Ruby, we're actually making changes that would support specifying revision version and avoiding breaks by loading the teeny version on demand if it's not cached for all tools in the cache. Supporting it takes some work and while we have already started, it will probably take a couple of months to roll it out. In the meantime, if you can, please avoid specifying the revision in the task.

**Comment by alepauly** [comment]
@StingyJack  - And to answer the issue of the documents saying one thing but the image having another: We rollout in stages, not all customers at once. When we start a rollout we update the readme file in a release branch ([example](https://github.com/actions/virtual-environments/blob/releases/ubuntu18/20200113/images/linux/Ubuntu1804-README.md)) and update the master version only after we've rolled out to all customers. We also start a new pre-release when the rollout begins, and then mark it as final when it completes.

We have plans in the backlog to improve this so you can see the image contents within the context of your build. Can't promise a delivery date but it's definitely something we know should be better.
